### PR TITLE
Expense services routes

### LINKED ISSUE #5: Backend: Build CRUD Endpoints for Expenses & Categories
### Task Description
Create the core API routes that the frontend will consume to manage expenses and categories.

### Requirements
- [ ] Implement endpoints for **Categories** (Create, Read list).
- [ ] Implement endpoints for **Expenses** (Create, Read list, Edit, Delete).
- [ ] Apply strict backend validation for all incoming JSON requests.
- [ ] Ensure all date formats strictly follow ISO 8601 (`YYYY-MM-DDTHH:mm:ss.sssZ`).
- [ ] Important: All queries must filter data by the authenticated `user_id`.

### PR DESCRIPTION
**Pull request by huzipro**
# Fixes #5

## Purpose
Implement and harden the service layer for Categories and Expenses, including cross-user data protection, standardized exception handling, and test coverage improvements requested during review.

## Changes
- Implemented/updated `CategoryService` and `ExpenseService` CRUD behavior with strict user ownership checks (`user_id`) in lookup/update/delete flows.
- Refactored service methods to reuse `get_by_id(...)` (DRY), removing repeated manual fetch patterns.
- Standardized exception strategy:
  - Introduced generic `EntityNotFoundException` and `DatabaseException` in `src/errors/main.py`.
  - Removed expense-specific exception file (`src/errors/expense.py`) to avoid exception explosion.
  - Added safe details nesting in `DatabaseException` context to avoid key collisions.
  - Added exception chaining (`raise ... from e`) and original error details for better debugging.
- Improved global FastAPI app exception handling:
  - Added structured logging for app exceptions (status, code, message, context, method, path, request_id).
  - Kept existing JSON error response contract.
- Updated type contracts/nitpicks from review:
  - Category service methods now accurately return `Category` (not `Optional`) where not-null by contract.
  - Removed redundant `self.db.add(category)` in update path for already managed entities.
  - `ExpenseService.get_all(...)` now returns `list[Expense]` and returns `[]` when no records exist.
  - `AppException.__init__` now uses `message: str | None = None`.
  - `ExpenseCreateSchema` no longer uses `from_attributes=True` (input schema).
- Test and fixture updates (including CodeRabbit/review nitpicks):
  - Added `src/domain/expense/tests/test_expense_service.py` with:
    - create uses authenticated user id
    - cross-user delete is blocked
    - empty list behavior for `get_all`
  - Expanded `src/domain/category/tests/test_category_service.py` with:
    - cross-user access/update/delete authorization checks
    - nonexistent category edge-case check
    - defensive fixture copy usage (`model_copy`) to avoid shared mutation
  - Cleaned fixture setup in `conftest.py`:
    - removed debug print
    - made `valid_category` synchronous fixture
    - removed redundant `user` dependency from `category` fixture
  - Removed redundant manual commit in category test where service already commits.

## How to Test
1. Ensure DB container is up:
   - `docker compose up -d`
2. Run quality checks:
   - `docker exec expenseapp-tools sh -lc "cd /app && uv run pre-commit run --all-files && uv run ruff check src conftest.py"`
3. Run tests:
   - `docker exec expenseapp-tools sh -lc "cd /app && uv run pytest -q"`
   - or focused:
     - `docker exec expenseapp-tools sh -lc "cd /app && uv run pytest src/domain/category/tests/test_category_service.py src/domain/expense/tests/test_expense_service.py -q"`

## ✅ Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

## Summary by CodeRabbit

* **New Features**
  * Added expense management API endpoints (create, list, view, update, delete)
  * Added category management functionality with full CRUD support
  * Implemented structured error responses with standardized error codes and messages

* **Tests**
  * Added comprehensive test coverage for category and expense operations with authorization validation

### CODE REVIEW

File: src/errors/main.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        self.context = context or {}
+
+        super().__init__(self.message)
+
+
+class ExpenseException(AppException):
+    pass
+
+
+class AuthExeption(AppException):
```
Comment (thawancomt) [inline src/errors/main.py]: This one could be removed as well

File: src/domain/expense/service.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    async def get_all(self, user_uid: UUID):
+        statement = select(Expense).where(Expense.user_id == user_uid)
+        result = await self.db.execute(statement)
+        expenses = result.scalars().all()
+        if not expenses:
+            raise NoExpenseFound(message="No Expense found for the user")
+        return expenses
+
+    # async def get_all_expenses(self) -> List[Expense]:
```
Comment (thawancomt) [inline src/domain/expense/service.py]: dead code shall be removed

File: src/domain/expense/service.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        try:
+            await self.db.commit()
+            await self.db.refresh(expense_orm)
+        except Exception:
+            await self.db.rollback()
+            raise ExpenseCreationDataBase()
+
+        return expense_orm
+
+    async def delete(self, object_id: Any, user_id: UUID) -> Optional[Expense]:
```
Comment (thawancomt) [inline src/domain/expense/service.py]: Be carefully about security, you can't delete a expense without check who is deleting it, ur not using the user_id...
This way any user can delete any expense for whoever.

Your query should ensure the object_id exist and belong to user that is deleting

**Review by thawancomt — CHANGES_REQUESTED** [review]
please review the following architectural adjustments for this PR:

1. Service Layer Reusability (DRY Principle)
The Service layer is repeating itself. Since you already implemented a secure get_by_id method (which handles the user_id validation), you do not need to manually fetch the object from the database in every other method.

Action: Refactor the update and delete methods to call await self.get_by_id(...) instead of rewriting the select statement.

2. Standardizing Exceptions (Exception Explosion)
We need to stop creating specific exceptions for every single entity (ExpenseNotFound, UserNotFound, etc.). At the end of the day, they are all the same error: a requested entity was not found in the database.

Action: Use a single, generic exception class. Whenever a search fails, raise it with context:
raise EntityNotFoundException(entity_name="Expense", object_id=1)

Why: This ensures our error logs are uniform and tells us exactly which domain failed and what ID caused the failure, making debugging infinitely easier.
